Provide more control over automatic generation of docstrings

stukowski opened the pull request:
I added a ``pybind11::docstring_options`` class, which allows controlling the generation of function signatures and docstrings by pybind11, either globally or selectively. 

The design of this class follows closely the [docstring_options](http://www.boost.org/doc/libs/1_61_0/libs/python/doc/html/reference/high_level_components/boost_python_docstring_options_h.html) facility of Boost.Python. 

Currently, pybind11 always generates a docstring for a function that consists of an auto-generated signature and an optional user-defined string. The new ``docstring_options`` class allows to suppress the generation of the signature. This is useful, for example, when one wants a function to have no docstring at all, because it should not show up in the Sphinx documentation.

Like its Boost.Python counterpart, the ``docstring_options`` class furthermore provides an option to suppress the registration of user-defined docstrings. This option suppresses the docstrings of modules, classes, and properties in addition to functions. (I have to admit, I'm not sure what the point of this option is. Why would one want to suppress their user-defined docstrings?)

aldanor commented:
This looks fairly reasonable if you need it, I guess.

I was thinking, one other neat thing to have would be to have a processor(s) for signatures and docstrings (e.g. so instead of having `very::long::ugly::c++::names` you could have`python.friendly.ones` and the generated sphinx docs could look nice out of the box; you can even make it generate proper xrefs etc). I kind of had to do it manually when I needed it but it'd be nice to have it in pybind11.

// Btw the tests fail on Python 2.7.

aldanor commented on include/pybind11/docstring_options.h:
```diff
@@ -0,0 +1,73 @@
+NAMESPACE_BEGIN(pybind11)
+
+class docstring_options
+{
```
Very minor, but: could you try using consistent braces indentation throughout the file? (there's at least three different ways here). 

This one is perhaps the most popular in the current codebase:

```cpp
foo {
    bar
}
```

aldanor commented on include/pybind11/docstring_options.h:
```diff
@@ -0,0 +1,73 @@
+        global_state().show_signatures = show_signatures;
+    }
+
+    ~docstring_options()
```
So what happens if it's heap-allocated and not destroyed at the end of the scope? Is it meant to never be created not on the stack?

jagerman replied:
> So what happens if it's heap-allocated and not destroyed at the end of the scope? Is it meant to never be created not on the stack?

It's basically just an RIAA-type class that overrides the globals while it exists then restores them when it gets destroyed.  So if it never gets destroyed, its override never gets restored (but this is basically true of any RAII class).  So I think the answer to the second question is "yes, usually", but I don't see that it would break if you managed it yourself on the heap (the worst case is that the values from before you instantiated it will never be restored).

aldanor replied:
Yea. You could `private` the `operator new` though to be safe :)

jagerman commented on include/pybind11/docstring_options.h:
```diff
@@ -0,0 +1,73 @@
+        return instance;
+    }
+
+    docstring_options(bool show_all)
```
This constructor needs an "explicit" (to prevent implicit conversion from bool).

jagerman commented on include/pybind11/docstring_options.h:
```diff
@@ -0,0 +1,73 @@
+        bool show_signatures;   //< Inclcude auto-generated function signatures in docstrings.
+    };
+
+    PYBIND11_NOINLINE inline static state &global_state() {
```
`PYBIND11_NOINLINE inline` isn't needed here.

Although perhaps you could just use a `static state global_state = { true, true };`, i.e. not protected behind a method?

stukowski commented:
I made some changes to the code to implement most of the helpful suggestions made above. 

The internal state structure is now private, and public getter functions have been added to the docstring_options class. I anticipate that this class can be extended in the future to give users additional control over the generation of docstrings, like e.g. the formatting of signatures as suggested by @aldanor.

jagerman commented:
The build failure is really minor (but simple to fix): you have a couple lines with trailing spaces being picked up by the style checks:

https://travis-ci.org/pybind/pybind11/jobs/174162279#L294

jagerman commented:
The last thing I think it needs is a bit of documentation (probably in `docs/advanced/functions.rst`) that describes how to use it and gives a quick example.

dean0x7d commented:
There's already some sphinx related stuff in the docs, so it would probably be good to keep those close together: http://pybind11.readthedocs.io/en/latest/advanced/misc.html#generating-documentation-using-sphinx

BTW Looks like there are tests missing for class and module docstrings. Although, is there a compelling reason for suppressing user-defined docstrings? (As opposed to signatures.)

stukowski commented:
I now added a brief introduction to the ``docstring_options`` class to the **Generating documentation using Sphinx** section of the manual.

I also changed the design of the class a bit. I didn't like the initialization constructor, which took two boolean flags to control the suppression of user-defined docstrings and auto-generated signatures. This is prone to errors and hard to remember. Explicit method calls are now necessary to configure the docstring generation (which is more verbose but easier to remember):

```
py::docstring_options doc_options;
doc_options.disable_signatures();
...
```

Also extended the test case to fully cover docstring_options' functionality.

jagerman commented:
Some other thoughts on the interface:

You could make the interface slightly nicer by making the `disable_*` and `enable_*` methods return the object, e.g.
```C++
docstring_options &enable_user_defined() { global_state().show_user_defined = true; return *this; }
```
which allows:
```C++
docstring_options doc_opts;
doc_opts.enable_user_defined().disable_signatures();
```

There's a second modification you could make to the interface that would make it less error-prone, but unfortunately can't be combined with the above.  The current code allows:
```C++
docstring_options().enable_user_defined();
```
which won't actually do anything useful.  You could prevent this by lvalue-qualifying the enable/disable methods, i.e.:
```C++
void enable_user_defined() & { global_state().show_user_defined = true; }
```
which makes
```C++
py::docstring_options().enable_user_defined();
```
a compile-time error.  Unfortunately, it also makes chaining (my first suggestion) a compile-time error.

aldanor commented:
Minor: I think you could also add this to indicate it's meant to be a stack-allocated thing.

```cpp
private: 
    void* operator new(size_t);
```

jagerman commented:
You could also avoid the potential calling-on-an-rvalue confusion by renaming it to something like `local_docstring_options` or `scoped_docstring_options` (and then you're free to add the chaining).

stukowski commented:
@jagerman: Thank you for your hint on how to prevent calling the enable/disable methods on a temporary by lvalue-qualifying the methods. This seems to work. But unlike you wrote, this doesn't infer with the chaining. With this definition
```c++
docstring_options& disable_user_defined() & { 
    global_state().show_user_defined = false; 
    return *this; 
}
```
I can do
```c++
py::docstring_options doc_options;
doc_options.disable_signatures().disable_user_defined();
```
But the following produces a compile-time error, exactly as we want:
```c++
py::docstring_options().disable_signatures();
```

@aldanor: I will make the ``new`` operator private as you suggest, but I am not fully convinced that it is necessary. Why would somebody try to allocate an instance of this class on the heap? Is this really something we need to prevent? The documentation clearly states that the settings will be in effect as long as the object exists. If somebody wants the settings to remain in effect forever by allocating the object on the heap, then be it so, I would say.

jagerman commented:
> But unlike you wrote, this doesn't infer with the chaining. 

Oh, indeed it does; I must have broken something else when I tried it.

jagerman commented:
> I will make the new operator private as you suggest, but I am not fully convinced that it is necessary.

I basically agree with you.  There is, however, some potential weirdness with the interplay of a stack-allocated and heap-allocated object, e.g.

```C++
docstring_options *b;
{
    docstring_options a;
    // set a options
    b = new docstring_options();
    // Set b options
}
// original options restored
delete b;
// Weirndess: "a" options are now in effect
```

The stack-only approach would make this impossible (since instantiation/destruction will then always be in FILO order).

stukowski commented:
@jagerman: Ok, I see what you mean. The new operator has now been deleted from the class definition to prevent this kind of wrong usage. Allocating instances on the heap using ``std::make_shared`` etc. is still possible though. So some loopholes remain.

aldanor commented:
We're not being *that* paranoid :) All in all, this lgtm to me.

aldanor commented:
(would prob make sense to squash on merge though)

wjakob commented:
I think that this is great but would suggest one more change. Can we call it just ``options`` instead of ``docstring_options``? It may be useful to be able to control other aspects of pybind11 with a RAII-style interface.

One thing that comes to mind would be automatic conversion of C++ ``camelCase`` function names to ``camel_case`` for a more pythonic feel.

aldanor commented:
Shouldn't the header be also renamed then?

stukowski commented:
I renamed the new docstring_options class to just 'options' to make it ready for future, more general uses. Its existing setter/getter methods have been renamed as well to make clear they are pertaining to generation of docstrings.

wjakob commented:
Beautiful, thank you!
